File: src/bin/day4.rs
//! Day 4: Ceres Search
//!
//! <https://adventofcode.com/2024/day/4>

use std::{collections::HashSet, fmt::Display};
use aoc_2024::Solution;

pub struct Day4;

static DIRECTIONS: [(i8, i8); 8] = [
    (0, 1),
    (0, -1),
    (1, 0),
    (-1, 0),
    (1, 1),
    (1, -1),
    (-1, 1),
    (-1, -1),
];
static XMAS: &[u8; 4] = b"XMAS";

lazy_static::lazy_static! {
    static ref CORNERS: HashSet<u8> = HashSet::from_iter([b'M', b'S']);
}

impl Day4 {
    #[inline]
    #[must_use]
    fn get_grid<T: Display>(inp: T) -> Vec<Vec<u8>> {
        inp.to_string()
            .lines()
            .map(|line| line.as_bytes().to_vec())
            .collect::<Vec<_>>()
    }

    #[must_use]
    fn find_xmas(
        grid: &[Vec<u8>],
        n_rows: usize,
        n_cols: usize,
        i: usize, j: usize,
        dr: usize, dc: usize,
        progression: usize,
    ) -> bool {
        if progression == 4 {
            return true;
        }

        if !(0..n_rows).contains(&i) || !(0..n_cols).contains(&j) {
            return false;
        }

        if grid[i][j] == XMAS[progression] {
            return Self::find_xmas(
                grid,
                n_rows, n_cols,
                i.wrapping_add(dr),
                j.wrapping_add(dc),
                dr, dc,
                progression + 1,
            )
        }
        false
    }
}

impl Solution for Day4 {
    const NAME: &'static str = "Ceres Search";

    fn part_one<T: Display>(&self, inp: T) -> usize {
        let grid = &Self::get_grid(inp);
        let n_rows = grid.len();
        let n_cols = grid[0].len();

        (0..n_rows)
            .flat_map(|i| (0..n_cols)
                .flat_map(move |j| DIRECTIONS
                    .iter()
                    .filter(move |&&(dr, dc)| Self::find_xmas(grid, n_rows, n_cols, i, j, dr as usize, dc as usize, 0))
                )
            )
            .count()
    }

    fn part_two<T: Display>(&self, inp: T) -> usize {
        let grid = &Self::get_grid(inp);
        let n_cols = grid[0].len();

        (1..grid.len() - 1)
            .flat_map(|i| (1..n_cols - 1)
                .filter(move |&j| {
                    if grid[i][j] != b'A' {
                        return false;
                    }

                    let diag1 = HashSet::from_iter([
                        grid[i - 1][j - 1],  grid[i + 1][j + 1]
                    ]);
                    let diag2 = HashSet::from_iter([
                        grid[i - 1][j + 1],  grid[i + 1][j - 1]
                    ]);
                    diag1 == *CORNERS && diag2 == *CORNERS
                })
            )
            .count()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 2578);
        assert_eq!(p2, 1972);
    }
}

fn main() {
    aoc_2024::run_day(4, &Day4);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}

File: aoc-py/solutions/day4.py
"""
Day 4: Ceres Search

https://adventofcode.com/2024/day/4
"""
__all__ = ('Day4',)

from typing import ClassVar

from ..solution import Solution

class Day4(Solution):
    NAME: ClassVar[str] = 'Ceres Search'
    DIRECTIONS: ClassVar[tuple[tuple[int, int], ...]] = (
        (0, 1),
        (0, -1),
        (1, 0),
        (-1, 0),
        (1, 1),
        (1, -1),
        (-1, 1),
        (-1, -1),
    )
    CORNERS: ClassVar[set[str]] = {'M', 'S'}

    def _find_xmas(
        self,
        grid: list[str],
        n_rows: int,
        n_cols: int,
        i: int, j: int,
        dr: int, dc: int,
        progression: int,
    ) -> bool:
        if progression == 4:
            return True

        if i not in range(n_rows) or j not in range(n_cols):
            return False

        if grid[i][j] == 'XMAS'[progression]:
            return self._find_xmas(grid, n_rows, n_cols, i + dr, j + dc, dr, dc, progression + 1)
        return False

    def part_one(self, inp: str) -> int:
        grid = inp.splitlines()
        n_rows = len(grid)
        n_cols = len(grid[0])

        total = 0

        for i, row in enumerate(grid):
            for j, letter in enumerate(row):
                # potential start of XMAS chain
                if letter == 'X':
                    for dr, dc in self.DIRECTIONS:
                        total += self._find_xmas(grid, n_rows, n_cols, i, j, dr, dc, 0)
        return total

    def part_two(self, inp: str) -> int:
        grid = inp.splitlines()
        n_cols = len(grid[0])

        total = 0

        for i in range(1, len(grid) - 1):
            for j in range(1, n_cols - 1):
                # potential center of MAS cross
                if grid[i][j] == 'A':
                    tl = grid[i - 1][j - 1]
                    tr = grid[i - 1][j + 1]
                    bl = grid[i + 1][j - 1]
                    br = grid[i + 1][j + 1]

                    total += {tl, br} == self.CORNERS and {tr, bl} == self.CORNERS
        return total

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 2578
        assert p2 == 1972